pkg/scheduler/plugins/gang/gang_cache.go

```diff
@@ -0,0 +1,198 @@
+package gang
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+
+	"k8s.io/api/core/v1"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/scheduler-plugins/apis/scheduling/v1alpha1"
+
+	"github.com/koordinator-sh/koordinator/apis/extension"
+	schedulingconfig "github.com/koordinator-sh/koordinator/apis/scheduling/config"
+)
+
+type gangCache struct {
+	lock       *sync.RWMutex
+	gangItems  map[string]*Gang
+	pluginArgs *schedulingconfig.GangSchedulingArgs
+}
+
+func NewGangCache() *gangCache {
+	return &gangCache{
+		gangItems:  make(map[string]*Gang),
+		lock:       new(sync.RWMutex),
+		pluginArgs: &schedulingconfig.GangSchedulingArgs{},
+	}
+}
+
+/*****************Cache BEG*****************/
+func (gangCache *gangCache) createGangToCache(gangId string) *Gang {
+	gangCache.lock.Lock()
+	defer gangCache.lock.Unlock()
+
+	if gang, exist := gangCache.gangItems[gangId]; exist {
+		return gang
+	}
+
+	gang := NewGang(gangId)
+	gangCache.gangItems[gangId] = gang
+	return gang
+}
+
+func (gangCache *gangCache) GetGangFromCache(gangId string) *Gang {
+	gangCache.lock.RLock()
+	defer gangCache.lock.RUnlock()
+
+	return gangCache.gangItems[gangId]
+}
+
+func (gangCache *gangCache) deleteGangFromCache(gangId string) {
+	gangCache.lock.Lock()
+	defer gangCache.lock.Unlock()
+
+	delete(gangCache.gangItems, gangId)
+}
+
+/*****************Cache END*****************/
+
+/******************Pod Event BEG******************/
+func (gangCache *gangCache) OnPodAdd(obj interface{}) {
+	pod, ok := obj.(*v1.Pod)
+	if !ok {
+		return
+	}
+
+	if !IsValidPod(pod) {
+		return
+	}
+	var gang *Gang
+	//the gang is created in either CRD way or Annotation way
+	if _, exist := pod.Labels[v1alpha1.PodGroupLabel]; exist {
+		gangName := pod.Labels[v1alpha1.PodGroupLabel]
+		gangNamespace := pod.Namespace
+		gangId := GetNamespaceSplicingName(gangNamespace, gangName)
+		gang = gangCache.GetGangFromCache(gangId)
```

Comment:
consider a time order, first receive pod, then receive pg; we should create gang first, and then init gang, to avoid lost pod event. please add a ut to cover this case

Reply:
yes, we should also create a gang here rather than return

Reply:
solved